File: base/OPDFO_220531/fsrc/classical/cobyla/cobyla.f90
module cobyla_mod
!--------------------------------------------------------------------------------------------------!
! Classical mode. Not maintained. Not recommended. Please use the modernized version instead.
!
! The usage is the same as the modernized version.
!--------------------------------------------------------------------------------------------------!

implicit none
private
public :: cobyla


contains


subroutine cobyla(calcfc, m, x, f, &
    & cstrv, constr, &
    & f0, constr0, &
    & nf, rhobeg, rhoend, ftarget, ctol, cweight, maxfun, iprint, &
    & eta1, eta2, gamma1, gamma2, xhist, fhist, chist, conhist, maxhist, maxfilt, info)

! Generic modules
use, non_intrinsic :: consts_mod, only : DEBUGGING
use, non_intrinsic :: consts_mod, only : MAXFUN_DIM_DFT, MAXFILT_DFT, IPRINT_DFT
use, non_intrinsic :: consts_mod, only : RHOBEG_DFT, RHOEND_DFT, CTOL_DFT, CWEIGHT_DFT, FTARGET_DFT
use, non_intrinsic :: consts_mod, only : RP, IK, ZERO, ONE, TWO, HALF, TEN, TENTH, EPS, MSGLEN
use, non_intrinsic :: debug_mod, only : assert, errstop, warning
use, non_intrinsic :: evaluate_mod, only : evaluate, moderatex
use, non_intrinsic :: history_mod, only : prehist
use, non_intrinsic :: infnan_mod, only : is_finite
use, non_intrinsic :: memory_mod, only : safealloc
use, non_intrinsic :: pintrf_mod, only : OBJCON
!use, non_intrinsic :: selectx_mod, only : isbetter
use, non_intrinsic :: preproc_mod, only : preproc

! Solver-specific modules
use, non_intrinsic :: cobylb_mod, only : cobylb

implicit none

! Compulsory arguments
procedure(OBJCON) :: calcfc
real(RP), intent(inout) :: x(:)
real(RP), intent(out) :: f
integer(IK), intent(in) :: m

! Optional inputs
integer(IK), intent(in), optional :: iprint
integer(IK), intent(in), optional :: maxfilt
integer(IK), intent(in), optional :: maxfun
integer(IK), intent(in), optional :: maxhist
real(RP), intent(in), optional :: constr0(:)
real(RP), intent(in), optional :: ctol
real(RP), intent(in), optional :: cweight
real(RP), intent(in), optional :: eta1
real(RP), intent(in), optional :: eta2
real(RP), intent(in), optional :: f0
real(RP), intent(in), optional :: ftarget
real(RP), intent(in), optional :: gamma1
real(RP), intent(in), optional :: gamma2
real(RP), intent(in), optional :: rhobeg
real(RP), intent(in), optional :: rhoend

! Optional outputs
integer(IK), intent(out), optional :: info
integer(IK), intent(out), optional :: nf
real(RP), intent(out), allocatable, optional :: chist(:)
real(RP), intent(out), allocatable, optional :: conhist(:, :)
real(RP), intent(out), allocatable, optional :: constr(:)
real(RP), intent(out), allocatable, optional :: fhist(:)
real(RP), intent(out), allocatable, optional :: xhist(:, :)
real(RP), intent(out), optional :: cstrv

! Local variables
character(len=*), parameter :: ifmt = '(I0)'  ! I0: use the minimum number of digits needed to print
character(len=*), parameter :: solver = 'COBYLA'
character(len=*), parameter :: srname = 'COBYLA'
character(len=MSGLEN) :: wmsg
integer(IK) :: info_loc
integer(IK) :: iprint_loc
integer(IK) :: maxfilt_loc
integer(IK) :: maxfun_loc
integer(IK) :: maxhist_loc
integer(IK) :: n
integer(IK) :: nf_loc
integer(IK) :: nhist
real(RP) :: cstrv_loc
real(RP) :: ctol_loc
real(RP) :: cweight_loc
real(RP) :: eta1_loc
real(RP) :: eta2_loc
real(RP) :: ftarget_loc
real(RP) :: gamma1_loc
real(RP) :: gamma2_loc
real(RP) :: rhobeg_loc
real(RP) :: rhoend_loc
real(RP), allocatable :: constr_loc(:)
real(RP), allocatable :: chist_loc(:)
real(RP), allocatable :: conhist_loc(:, :)
real(RP), allocatable :: fhist_loc(:)
real(RP), allocatable :: xhist_loc(:, :)

! Preconditions
if (DEBUGGING) then
    call assert(present(f0) .eqv. present(constr0), 'F0 and CONSTR0 are both present or both absent', srname)
end if

! Sizes
n = int(size(x), kind(n))

! Exit if the size of CONSTR0 is inconsistent with M.
if (present(constr0)) then
    if (size(constr0) /= m) then
        if (DEBUGGING) then
            call errstop(srname, 'SIZE(CONSTR0) /= M. Exiting')
        else
            call warning(srname, 'SIZE(CONSTR0) /= M. Exiting')
            return
        end if
    end if
end if

! Allocate memory for CONSTR_LOC, since M is now available.
call safealloc(constr_loc, m)  ! NOT removable even in F2003!

!! If the user provides the function & constraint value at X0, then set up F_X0 and CONSTR_X0.
!if (present(f0) .and. present(constr0)) then
!    fc_x0_provided = .true.
!    !--------------------------------------------------!
!    call safealloc(x0, n)  ! Removable in F2003.
!    call safealloc(constr_x0, m)  ! Removable in F2003.
!    !--------------------------------------------------!
!    x0 = x
!    f_x0 = f0
!    constr_x0 = constr0
!else
!    call evaluate(calcfc, x, f_x0, constr_x0)
!end if

if (present(f0) .and. present(constr0) .and. all(is_finite(x))) then
    f = f0
    constr_loc = constr0
    cstrv_loc = maxval([ZERO, -constr_loc])
else
    ! Replace any NaN in X by ZERO and Inf/-Inf in X by HUGENUM/-HUGENUM.
    x = moderatex(x)
    call evaluate(calcfc, x, f, constr_loc, cstrv_loc)
end if

! If RHOBEG is present, then RHOBEG_LOC is a copy of RHOBEG; otherwise, RHOBEG_LOC takes the default
! value for RHOBEG, taking the value of RHOEND into account. Note that RHOEND is considered only if
! it is present and it is VALID (i.e., finite and positive). The other inputs are read similarly.
if (present(rhobeg)) then
    rhobeg_loc = rhobeg
elseif (present(rhoend)) then
    ! Fortran does not take short-circuit evaluation of logic expressions. Thus it is WRONG to
    ! combine the evaluation of PRESENT(RHOEND) and the evaluation of IS_FINITE(RHOEND) as
    ! "IF (PRESENT(RHOEND) .AND. IS_FINITE(RHOEND))". The compiler may choose to evaluate the
    ! IS_FINITE(RHOEND) even if PRESENT(RHOEND) is false!
    if (is_finite(rhoend) .and. rhoend > ZERO) then
        rhobeg_loc = max(TEN * rhoend, RHOBEG_DFT)
    else
        rhobeg_loc = RHOBEG_DFT
    end if
else
    rhobeg_loc = RHOBEG_DFT
end if

if (present(rhoend)) then
    rhoend_loc = rhoend
elseif (rhobeg_loc > 0) then
    rhoend_loc = max(EPS, min(TENTH * rhobeg_loc, RHOEND_DFT))
else
    rhoend_loc = RHOEND_DFT
end if

if (present(ctol)) then
    ctol_loc = ctol
else
    ctol_loc = CTOL_DFT
end if

if (present(cweight)) then
    cweight_loc = cweight
else
    cweight_loc = CWEIGHT_DFT
end if

if (present(ftarget)) then
    ftarget_loc = ftarget
else
    ftarget_loc = FTARGET_DFT
end if

if (present(maxfun)) then
    maxfun_loc = maxfun
else
    maxfun_loc = MAXFUN_DIM_DFT * n
end if

if (present(iprint)) then
    iprint_loc = iprint
else
    iprint_loc = IPRINT_DFT
end if

if (present(eta1)) then
    eta1_loc = eta1
elseif (present(eta2)) then
    if (eta2 > ZERO .and. eta2 < ONE) then
        eta1_loc = max(EPS, eta2 / 7.0_RP)
    end if
else
    eta1_loc = TENTH
end if

if (present(eta2)) then
    eta2_loc = eta2
elseif (eta1_loc > ZERO .and. eta1_loc < ONE) then
    eta2_loc = (eta1_loc + TWO) / 3.0_RP
else
    eta2_loc = 0.7_RP
end if

if (present(gamma1)) then
    gamma1_loc = gamma1
else
    gamma1_loc = HALF
end if

if (present(gamma2)) then
    gamma2_loc = gamma2
else
    gamma2_loc = TWO
end if

if (present(maxhist)) then
    maxhist_loc = maxhist
else
    maxhist_loc = maxval([maxfun_loc, n + 2_IK, MAXFUN_DIM_DFT * n])
end if

if (present(maxfilt)) then
    maxfilt_loc = maxfilt
else
    maxfilt_loc = MAXFILT_DFT
end if

! Preprocess the inputs in case some of them are invalid. It does nothing if all inputs are valid.
call preproc(solver, n, iprint_loc, maxfun_loc, maxhist_loc, ftarget_loc, rhobeg_loc, rhoend_loc, &
    & m=m, ctol=ctol_loc, cweight=cweight_loc, eta1=eta1_loc, eta2=eta2_loc, gamma1=gamma1_loc, &
    & gamma2=gamma2_loc, maxfilt=maxfilt_loc)

! Further revise MAXHIST_LOC according to MAXMEMORY, and allocate memory for the history.
! In MATLAB/Python/Julia/R implementation, we should simply set MAXHIST = MAXFUN and initialize
! CHIST = NaN(1, MAXFUN), CONHIST = NaN(M, MAXFUN), FHIST = NaN(1, MAXFUN), XHIST = NaN(N, MAXFUN)
! if they are requested; replace MAXFUN with 0 for the history that is not requested.
call prehist(maxhist_loc, n, present(xhist), xhist_loc, present(fhist), fhist_loc, &
    & present(chist), chist_loc, m, present(conhist), conhist_loc)

!--------------------------------------------------------------------------------------------------!
!-------------------- Call COBYLB, which performs the real calculations. --------------------------!
!!!! ETA1, ETA2, GAMAA1, GAMMA2, MAXFILT, CTOL, CWEIGHT are not used in the classical mode. !!!!
call cobylb(calcfc, iprint_loc, maxfun_loc, rhobeg_loc, rhoend_loc, constr_loc, x, cstrv_loc, f, info_loc, &
    & nf_loc, xhist_loc, fhist_loc, chist_loc, conhist_loc)
!--------------------------------------------------------------------------------------------------!
!--------------------------------------------------------------------------------------------------!

! Write the outputs.

! Copy CONSTR_LOC to CONSTR if needed.
if (present(constr)) then
    !--------------------------------------------------!
    call safealloc(constr, m)  ! Removable in F2003.
    !--------------------------------------------------!
    constr = constr_loc
end if
deallocate (constr_loc)

if (present(cstrv)) then
    cstrv = cstrv_loc
end if

if (present(nf)) then
    nf = nf_loc
end if

if (present(info)) then
    info = info_loc
end if

! Copy XHIST_LOC to XHIST if needed.
if (present(xhist)) then
    nhist = min(nf_loc, int(size(xhist_loc, 2), IK))
    !----------------------------------------------------!
    call safealloc(xhist, n, nhist)  ! Removable in F2003.
    !----------------------------------------------------!
    xhist = xhist_loc(:, 1:nhist)
    ! N.B.:
    ! 0. Allocate XHIST as long as it is present, even if the size is 0; otherwise, it will be
    ! illegal to enquire XHIST after exit.
    ! 1. Even though Fortran 2003 supports automatic (re)allocation of allocatable arrays upon
    ! intrinsic assignment, we keep the line of SAFEALLOC, because some very new compilers (Absoft
    ! Fortran 21.0) are still not standard-compliant in this respect.
    ! 2. NF may not be present. Hence we should NOT use NF but NF_LOC.
    ! 3. When SIZE(XHIST_LOC, 2) > NF_LOC, which is the normal case in practice, XHIST_LOC contains
    ! GARBAGE in XHIST_LOC(:, NF_LOC + 1 : END). Therefore, we MUST cap XHIST at NF_LOC so that
    ! XHIST cointains only valid history. For this reason, there is no way to avoid allocating
    ! two copies of memory for XHIST unless we declare it to be a POINTER instead of ALLOCATABLE.
end if
! F2003 automatically deallocate local ALLOCATABLE variables at exit, yet we prefer to deallocate
! them immediately when they finish their jobs.
deallocate (xhist_loc)

! Copy FHIST_LOC to FHIST if needed.
if (present(fhist)) then
    nhist = min(nf_loc, int(size(fhist_loc), IK))
    !--------------------------------------------------!
    call safealloc(fhist, nhist)  ! Removable in F2003.
    !--------------------------------------------------!
    fhist = fhist_loc(1:nhist)  ! The same as XHIST, we must cap FHIST at NF_LOC.
end if
deallocate (fhist_loc)

! Copy CHIST_LOC to CHIST if needed.
if (present(chist)) then
    nhist = min(nf_loc, int(size(chist_loc), IK))
    !--------------------------------------------------!
    call safealloc(chist, nhist)  ! Removable in F2003.
    !--------------------------------------------------!
    chist = chist_loc(1:nhist)  ! The same as XHIST, we must cap CHIST at NF_LOC.
end if
deallocate (chist_loc)

! Copy CONHIST_LOC to CONHIST if needed.
if (present(conhist)) then
    nhist = min(nf_loc, int(size(conhist_loc, 2), IK))
    !----------------------------------------------------------!
    call safealloc(conhist, m, nhist)  ! Removable in F2003.
    !----------------------------------------------------------!
    conhist = conhist_loc(:, 1:nhist)  ! The same as XHIST, we must cap CONHIST at NF_LOC.
end if
deallocate (conhist_loc)

! If NF_LOC > MAXHIST_LOC, warn that not all history is recorded.
if ((present(xhist) .or. present(fhist) .or. present(chist) .or. present(conhist)) .and. maxhist_loc < nf_loc) then
    write (wmsg, ifmt) maxhist_loc
    call warning(solver, 'Only the history of the last '//trim(wmsg)//' iteration(s) is recorded')
end if

end subroutine cobyla


end module cobyla_mod

File: python/src/prima/cobyla/cobyla.py
from prima.common.evaluate import evaluate, moderatex
from prima.common.consts import EPS, RHOBEG_DEFAULT, RHOEND_DEFAULT, \
    CTOL_DEFAULT, CWEIGHT_DEFAULT, FTARGET_DEFAULT, IPRINT_DEFAULT, \
    MAXFUN_DIM_DEFAULT, DEBUGGING
from prima.common.preproc import preproc
from prima.common.present import present
from prima.cobyla.cobylb import cobylb
import numpy as np
from dataclasses import dataclass


# Return type for COBYLA
@dataclass
class COBYLAResult:
    x: np.ndarray
    f: float
    constr: np.ndarray
    cstrv: float
    nf: int
    xhist: np.ndarray | None
    fhist: np.ndarray | None
    chist: np.ndarray | None
    conhist: np.ndarray | None
    info: int


def cobyla(calcfc, num_constraints, x, f0=None, constr0=None, rhobeg=None,
           rhoend=None, ftarget=FTARGET_DEFAULT, ctol=CTOL_DEFAULT, cweight=CWEIGHT_DEFAULT,
           maxfun=None, iprint=IPRINT_DEFAULT, eta1=None, eta2=None, gamma1=0.5, gamma2=2,
           maxhist=None, maxfilt=2000):
    """
    Among all the arguments, only CALCFC, M, and X are obligatory. The others are OPTIONAL and you
    can neglect them unless you are familiar with the algorithm. Any unspecified optional input will
    take the default value detailed below. For instance, we may invoke the solver as follows.

    First define CALCFC, NUM_CONSTRAINTS, and X, and then do the following.
    result = cobyla(calcfc, num_constraints, x)

    or

    First define CALCFC, NUM_CONSTRAINTS, and X, and then do the following.
    result = cobyla(calcfc, num_constraints, x, rhobeg=0.5, rhoend=1E-3, maxfun=100)

    IMPORTANT NOTICE: The user must set NUM_CONSTRAINTS correctly to the number of constraints, namely
    the size of CONSTR introduced below. Set NUM_CONSTRAINTS to 0 if there is no constraint.

    See examples/cobyla_example.py for a concrete example.

    A detailed introduction to the arguments is as follows. The section below is largely similar to the
    comments in the Fortran version but has been slightly modified to reflect the differences in
    some of the arguments.

    CALCFC
    F, CONSTR = CALCFC(X) should evaluate the objective function and constraints at the given
    vector X; it should return the objective function value and the constraint value. It must be
    provided by the user, and its definition must conform to the following interface:
    !-------------------------------------------------------------------------!
        def calcfc(x):
            # ... calculations ...
            return f, constr
    !-------------------------------------------------------------------------!
    X, F, and CONSTR are numpy arrays. If there are no constraints, CONSTR is expected to be an
    empty numpy array.

    NUM_CONSTRAINTS
    NUM_CONSTRAINTS must be set to the number of constraints
    N.B.:
    1. Why don't we define NUM_CONSTRAINTS as optional and determine it from running CALCFC?
    Because we need to know it ahead of time to setup up various matrices and vectors. Since CALCFC
    might be expensive, we prefer to avoid running it unnecessarily.

    X
    X should be an N-dimensional vector that contains the starting point, N being the
    dimension of the problem.

    F0
    F0, if present, should be set to the objective function value of the starting X.

    CONSTR0
    CONSTR0, if present, should be set to the constraint value of the starting X; in addition,
    SIZE(CONSTR0) must be NUM_CONSTRAINTS, or the solver will abort.

    RHOBEG, RHOEND
    Default: RHOBEG = 1, RHOEND = 10^-6. RHOBEG and RHOEND must be set to
    the initial and final values of a trust-region radius, both being positive and RHOEND <= RHOBEG.
    Typically RHOBEG should be about one tenth of the greatest expected change to a variable, and
    RHOEND should indicate the accuracy that is required in the final values of the variables.

    FTARGET
    Default: -Inf.
    FTARGET is the target function value. The algorithm will terminate when a feasible point with a
    function value <= FTARGET is found.

    CTOL
    Default: machine epsilon.
    CTOL is the tolerance of constraint violation. Any X with MAXVAL(-CONSTR(X)) <= CTOL is
    considered feasible.
    N.B.: 1. CTOL is absolute, not relative. 2. CTOL is used only when selecting the returned X.
    It does not affect the iterations of the algorithm.

    CWEIGHT
    Default: CWEIGHT_DFT defined in common/consts.py.
    CWEIGHT is the weight that the constraint violation takes in the selection of the returned X.

    MAXFUN
    Default: MAXFUN_DIM_DFT*N with MAXFUN_DIM_DFT defined in common/consts.py.
    MAXFUN is the maximal number of calls of CALCFC.

    IPRINT
    Default: 0.
    The value of IPRINT should be set to 0, 1, -1, 2, -2, 3, or -3, which controls how much
    information will be printed during the computation:
    0: there will be no printing;
    1: a message will be printed to the screen at the return, showing the best vector of variables
        found and its objective function value;
    2: in addition to 1, each new value of RHO is printed to the screen, with the best vector of
        variables so far and its objective function value; each new value of CPEN is also printed;
    3: in addition to 2, each function evaluation with its variables will be printed to the screen;
    -1, -2, -3: the same information as 1, 2, 3 will be printed, not to the screen but to a file
        named COBYLA_output.txt; the file will be created if it does not exist; the new output will
        be appended to the end of this file if it already exists.
    Note that IPRINT = +/-3 can be costly in terms of time and/or space.

    ETA1, ETA2, GAMMA1, GAMMA2
    Default: ETA1 = 0.1, ETA2 = 0.7, GAMMA1 = 0.5, and GAMMA2 = 2.
    ETA1, ETA2, GAMMA1, and GAMMA2 are parameters in the updating scheme of the trust-region radius
    detailed in the subroutine TRRAD in trustregion.py. Roughly speaking, the trust-region radius
    is contracted by a factor of GAMMA1 when the reduction ratio is below ETA1, and enlarged by a
    factor of GAMMA2 when the reduction ratio is above ETA2. It is required that 0 < ETA1 <= ETA2
    < 1 and 0 < GAMMA1 < 1 < GAMMA2. Normally, ETA1 <= 0.25. It is NOT advised to set ETA1 >= 0.5.

    XHIST, FHIST, CHIST, CONHIST, MAXHIST
    MAXHIST: Input, INTEGER(IK) scalar, default: MAXFUN
    XHIST, if present, will output the history of iterates; FHIST, if present, will output the
    history function values; CHIST, if present, will output the history of constraint violations;
    CONHIST, if present, will output the history of constraint values; MAXHIST should be a
    nonnegative integer, and XHIST/FHIST/CHIST/CONHIST will output only the history of the last
    MAXHIST iterations. Therefore, MAXHIST= 0 means XHIST/FHIST/CONHIST/CHIST will output nothing,
    while setting MAXHIST = MAXFUN requests XHIST/FHIST/CHIST/CONHIST to output all the history.
    If XHIST is present, its size at exit will be (N, min(NF, MAXHIST)); if FHIST is present, its
    size at exit will be min(NF, MAXHIST); if CHIST is present, its size at exit will be
    min(NF, MAXHIST); if CONHIST is present, its size at exit will be (M, min(NF, MAXHIST)).

    IMPORTANT NOTICE:
    Setting MAXHIST to a large value can be costly in terms of memory for large problems.
    MAXHIST will be reset to a smaller value if the memory needed exceeds MAXHISTMEM defined in
    CONSTS_MOD (see consts.F90 under the directory named "common").
    Use *HIST with caution!!! (N.B.: the algorithm is NOT designed for large problems).

    MAXFILT
    Input, INTEGER(IK) scalar.
    MAXFILT is a nonnegative integer indicating the maximal length of the filter used for selecting
    the returned solution; default: MAXFILT_DFT (a value lower than MIN_MAXFILT is not recommended);
    see common/consts.F90 for the definitions of MAXFILT_DFT and MIN_MAXFILT.

    Output: A Python dataclass with the following fields
    X
    X will be set to the best vector of variables found.

    F
    F will be set to the objective function value at X.

    CSTRV
    Optional, only appears if CSTRV is True. CSTRV will be set to the constraint violation of X at
    exit, i.e., MAXVAL([-CONSTR(X), 0]).

    CONSTR
    Optional, only appears if CONSTR is True. CONSTR will be set to the constraint value of X at
    exit.

    NF
    NF will be set to the number of calls of CALCFC at exit.

    INFO
    INFO is the exit flag. It will be set to one of the following values defined in the module
    INFOS_MOD (see common/infos.f90):
    SMALL_TR_RADIUS: the lower bound for the trust region radius is reached;
    FTARGET_ACHIEVED: the target function value is reached;
    MAXFUN_REACHED: the objective function has been evaluated MAXFUN times;
    MAXTR_REACHED: the trust region iteration has been performed MAXTR times (MAXTR = 2*MAXFUN);
    NAN_INF_X: NaN or Inf occurs in X;
    DAMAGING_ROUNDING: rounding errors are becoming damaging.
    !--------------------------------------------------------------------------!
    The following case(s) should NEVER occur unless there is a bug.
    NAN_INF_F: the objective function returns NaN or +Inf;
    NAN_INF_MODEL: NaN or Inf occurs in the model;
    TRSUBP_FAILED: a trust region step failed to reduce the model
    !--------------------------------------------------------------------------!
    """

    # Local variables
    solver = 'COBYLA'

    # Preconditions
    if DEBUGGING:
        assert (f0 is None) == (constr0 is None), "f0 and constr0 must be both present or both absent"

    # Sizes
    num_vars = len(x)

    # Exit if constr0 is not consistent with num_constraints
    if present(constr0):
        assert np.size(constr0) == num_constraints


    # If the user provides the function & constraint value at X0, then use them.
    # Otherwise, set [F0, CONSTR0] = [F(X0), CONSTR(X0)], so that COBYLB only needs a single interface.
    if (f0 is None and constr0 is None) or not all(np.isfinite(x)):
        # Replace any NaN in X by ZERO and Inf/-Inf in X by REALMAX/-REALMAX.
        x = moderatex(x)
        f0, constr0, cstrv0 = evaluate(calcfc, x)
    
    if present(rhobeg):
        rhobeg = rhobeg
    elif present(rhoend) and np.isfinite(rhoend) and rhoend > 0:
        rhobeg = np.maximum(10*rhoend, RHOBEG_DEFAULT)
    else:
        rhobeg = RHOBEG_DEFAULT

    if present(rhoend):
        rhoend = rhoend
    elif rhobeg > 0:
        rhoend = np.maximum(EPS, np.minimum(0.1*rhobeg, RHOEND_DEFAULT))
    else:
        rhoend = RHOEND_DEFAULT

    maxfun = maxfun if present(maxfun) else MAXFUN_DIM_DEFAULT*num_vars

    if present(eta1):
        eta1 = eta1
    elif present(eta2) and 0 < eta2 < 1:
        eta1 = np.maximum(EPS, eta2/7)
    else:
        eta1 = 0.1

    if present(eta2):
        eta2 = eta2
    else:
        if 0 < eta1 < 1:
            eta2 = (eta1 + 2)/3
        else:
            eta2 = 0.7

    maxhist = maxhist if present(maxhist) else np.max([maxfun, num_vars+2, MAXFUN_DIM_DEFAULT*num_vars])

    # Preprocess the inputs in case some of them are invalid. It does nothing if all inputs are valid.
    # Python: The underscores refer to `npt` and `x0`, neither of which are relevant here.
    iprint, maxfun, maxhist, ftarget, rhobeg, rhoend, _, maxfilt, ctol, cweight, \
        eta1, eta2, gamma1, gamma2, _ = preproc(solver, num_vars, iprint, maxfun, maxhist, \
                                                 ftarget, rhobeg, rhoend, num_constraints=num_constraints, ctol=ctol, cweight=cweight, \
                                                 eta1=eta1, eta2=eta2, gamma1=gamma1, gamma2=gamma2, maxfilt=maxfilt)

    
    # Further revise MAXHIST according to MAXHISTMEM, and allocate memory for the history.
    # In MATLAB/Python/Julia/R implementation, we should simply set MAXHIST = MAXFUN and initialize
    # CHIST = NaN(1, MAXFUN), CONHIST = NaN(M, MAXFUN), FHIST = NaN(1, MAXFUN), XHIST = NaN(N, MAXFUN)
    # if they are requested; replace MAXFUN with 0 for the history that is not requested.
    # prehist(maxhist, num_vars, present(xhist), xhist_loc, present(fhist), fhist_loc, &
    #     & present(chist), chist_loc, m, present(conhist), conhist_loc)
    

    # call coblyb, which performs the real calculations
    x, f, constr, cstrv, nf, xhist, fhist, chist, conhist, info = cobylb(calcfc, iprint, maxfilt, maxfun, ctol, cweight, eta1, eta2,
      ftarget, gamma1, gamma2, rhobeg, rhoend, constr0, f0, x, 
      cstrv0, maxhist)
    
    return COBYLAResult(x, f, constr, cstrv, nf, xhist, fhist, chist, conhist, info)
